### ISSUE TITLE
Graphical Modeler keeping track of which tools were already run

### ISSUE THREAD

**Issue by lukaszkruk** [qgis/QGIS#39521]
**Feature description.**

QGIS would rock even more if the Graphical Modeler kept track of which tools are run and have generated outputs, and allow to only run parts of the model.

**Additional context**
This would save a lot of time if, for example, I have a long processing pipeline with multiple tools that take long time to run, and for some reason have to change a tool parameter, or add another step somewhere towards the end of the pipeline. As things are now, the whole model needs to be re-run. It would be great if I could re-run only tools that are after the parameter change.

For what it's worth, this would make it more similar to ArcMap's Model Builder.

**Comment by gioman** [comment]
> QGIS would rock even more if the Graphical Modeler kept track of which tools are run and have generated outputs

@lukaszkruk agree. Note for example that GRASS do this (not only in the context of the modeller, but for any single tool in general) for any output: data/time created, tool used to generate, inputs and parameters.

**Comment by alexbruy** [comment]
Processing Modeler is a tool for creating non-interative models for complex workflows, not tool for running them in interactive mode.

**Comment by FelisPimeja** [comment]
> Processing Modeler is a tool for creating non-interative models for complex workflows, not tool for running them in interactive mode.

Why not? That could be a great feature. Especially for debugging while building a complex model. I really miss this functionality since using the same in FME.

**Comment by lukaszkruk** [comment]
> Processing Modeler is a tool for creating non-interative models for complex workflows, not tool for running them in interactive mode.

Yes, I believe you are right. I'm wondering if it is because

a) there was a conscious, deliberate decision made for the modeler to be non-interactive, or 
b) the non-interactive modeler was easier or faster to implement, but it could be expanded in the future to also work interactively

My feature request hopes it's the latter, and that interactivity could be added so that QGIS rocks even more.

**Comment by lukaszkruk** [comment]
@alexbruy wondering if you could respond to my question in the last post if you have a moment? Cheers!

**Comment by baswein** [comment]
I believe this is has been fixed with the "Run from Here" and "Run Selected" features.
